drivers/sdhc/rcar_mmc.c

```diff
@@ -0,0 +1,2388 @@
+	reg |= RCAR_MMC_OPTION_WIDTH_MASK | 0xEE;
+	rcar_mmc_write_reg32(dev, RCAR_MMC_OPTION, reg);
+
+	/* block count enable */
+	rcar_mmc_write_reg32(dev, RCAR_MMC_STOP, RCAR_MMC_STOP_SEC);
+	/* number of transfer blocks */
+	rcar_mmc_write_reg32(dev, RCAR_MMC_SECCNT, 0);
+
+	/*
+	 * SD_BUF0 data swap disabled.
+	 * Read/write access to SD_BUF0 can be performed with the 64-bit access.
+	 *
+	 * Note: when using the DMA, the bus width should be fixed at 64 bits.
+	 */
+	rcar_mmc_write_reg32(dev, RCAR_MMC_HOST_MODE, 0);
+	data->width_access_sd_buf0 = 8;
+
+	/* disable sampling clock controller, it is used for uhs/sdr104, hs200 and hs400 */
+	ret = rcar_mmc_disable_scc(dev);
+	if (ret) {
+		return ret;
+	}
+
+	/*
+	 * configure divider inside MMC controller
+	 * set maximum possible divider
+	 */
+	ios.clock = data->props.f_min;
+	rcar_mmc_set_clk_rate(dev, &ios);
+
+	data->restore_cfg_after_reset = 1;
+
+	return 0;
+}
+
+#ifdef CONFIG_RCAR_MMC_DMA_IRQ_DRIVEN_SUPPORT
+/**
+ * @brief MMC IRQ handler
+ *
+ * @param arg The pointer to an MMC device structure
+ *
+ * @retval none
+ */
+static void rcar_mmc_irq_handler(const void *arg)
+{
+	const struct device *dev = arg;
+
+	uint32_t dma_info1 = rcar_mmc_read_reg32(dev, RCAR_MMC_DMA_INFO1);
+	uint32_t dma_info2 = rcar_mmc_read_reg32(dev, RCAR_MMC_DMA_INFO2);
+
+	if (dma_info1 || dma_info2) {
+		struct mmc_rcar_data *data = dev->data;
+
+		rcar_mmc_write_reg32(dev, RCAR_MMC_DMA_INFO1_MASK, 0xfffffeff);
+		rcar_mmc_write_reg32(dev, RCAR_MMC_DMA_INFO2_MASK, ~0);
+		k_sem_give(&data->irq_xref_fin);
+	} else {
+		LOG_WRN("%s: warning: non-dma event triggers irq", dev->name);
+	}
+}
+#endif /* CONFIG_RCAR_MMC_DMA_IRQ_DRIVEN_SUPPORT */
+
+/**
+ * @brief Initialize and configure the Renesas MMC driver
+ *
+ * @param dev The pointer to an MMC device structure
+ *
+ * @retval 0 on success, negative errno code otherwise
+ */
+static int rcar_mmc_init(const struct device *dev)
+{
+	int ret = 0;
+	struct mmc_rcar_data *data = dev->data;
+	const struct mmc_rcar_cfg *cfg = dev->config;
+
+	if (unlikely(cfg == NULL)) {
+		LOG_ERR("%s: there isn't config inside device structure", dev->name);
+		return -EFAULT;
+	}
+
+#ifdef CONFIG_RCAR_MMC_DMA_IRQ_DRIVEN_SUPPORT
+	ret = k_sem_init(&data->irq_xref_fin, 0, 1);
+	if (ret) {
+		LOG_ERR("%s: can't init semaphore", dev->name);
+		return ret;
+	}
+#endif
+
+	if (unlikely(dev->data == NULL)) {
+		LOG_ERR("%s: there isn't data instance inside device structure", dev->name);
+		return -EFAULT;
+	}
+
+	DEVICE_MMIO_MAP(dev, K_MEM_CACHE_NONE);
+
+	/* Configure dt provided device signals when available */
+	ret = pinctrl_apply_state(cfg->pcfg, PINCTRL_STATE_DEFAULT);
+	if (ret < 0) {
+		LOG_ERR("%s: error can't apply pinctrl state", dev->name);
+		goto exit_unmap;
+	}
+
+	if (!device_is_ready(cfg->cpg_dev)) {
+		LOG_ERR("%s: error cpg_dev isn't ready", dev->name);
+		ret = -ENODEV;
+		goto exit_unmap;
+	}
+
+	ret = rcar_mmc_init_start_clk(cfg);
```

Comment:
In other SD drivers, the clock start is not run during device init. Rather, it is done within the `sdhc_request` implementation, typically based on CMD0 being sent. The reasoning here is that the clock init should be run after the SD card power is toggled, and the SD subsystem may call `sd_init_io` multiple times if the card fails init on the first attempt.

This is useful for cases such as a warm reset, where power may not have been toggled to the card, and the card may not be in an initialized state

Reply:
> In other SD drivers, the clock start is not run during device init.

Maybe I missed something, but let's take a look at these call chains:
```
imx_usdhc_init --> USDHC_Init --> CLOCK_EnableClock
```
and
```
mcux_sdif_init --> SDIF_Init --> CLOCK_EnableClock
``` 

In the case of my driver, this is the clock supply for the SD/MMC controller and the SD clock supply is controlled through the registers of the SD/MMC controller. As far as I understand, the same story applies to `nxp,lpc-sdif` and `nxp,imx-usdhc` as well.

-----
You are right, I leave the SD clock turned on at the end of the function  `rcar_mmc_init_controller_regs`.

Note, this controller supports the automatic disabling of SD clock signaling if there are no commands to SD/MMC. However, we have not enabled it yet, it requires some testing.

Reply:
Sorry, I am referring specifically to the SDMMC init sequence here (which is separate from clocking the SD host controller on all the controllers I have encountered). The SD spec requires sending 74 clock cycles to the SD card before CMD0 can be sent. This should either be done in `sdhc_set_io` when the clock is enabled, or in `sdhc_request` when CMD0 is sent.

Reply:
This only performs the delay. The spec itself requires that the controller send 74 clock cycles to the card before any commands can be sent. The USDHC and LPC SDIF controllers both also automatically disable clocks when not sending commands. However, they expose a register to trigger sending 74 clock cycles specifically for this portion of the specification:

![image](https://github.com/zephyrproject-rtos/zephyr/assets/34665051/a12270da-5179-40c4-96c8-a338c9bc1a51)

If you do not have such a register on your hardware, you can also likely force the clock to be active for a millisecond or so to satisfy this requirement. Is there somewhere in this driver where this occurs that I am missing?